Add BFCM promo

bor0 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Implement BFCM in-app promo.

⚠️ After merging, make sure to update `wccom_iam_promos` option on WCCOM with the necessary data (see [this](21881-gh-Automattic/woocommerce.com)). ⚠️

Closes 21814-gh-Automattic/woocommerce.com.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Backward compatibility I

1. Switch to `trunk`
2. Ensure the [Marketplace endpoint](https://github.com/woocommerce/woocommerce/blob/f9cf25e/plugins/woocommerce/includes/admin/class-wc-admin-assets.php#L562) returns BFCM data (i.e. apply [this patch](https://github.com/user-attachments/files/17175811/test.patch))
3. Open the following screens, make sure the frontend code still works:
    - `WooCommerce > Extensions` (make sure you see the weekly deals notice (if you don't, run `localStorage.removeItem( 'wc-marketplaceNoticeClosed-promotion-0' )`))
    - `WooCommerce > Home` (make sure screen renders)
    - `Marketing > Overview` (make sure screen renders)
    - `Marketing > Coupons` (make sure screen renders)

#### Backward compatibility II

1. Switch to this branch
2. Ensure the [Marketplace endpoint](https://github.com/woocommerce/woocommerce/blob/f9cf25e/plugins/woocommerce/includes/admin/class-wc-admin-assets.php#L562) doesn't return BFCM data (i.e. the data in [this patch](https://github.com/user-attachments/files/17175811/test.patch) is **not** returned)
3. Open the following screens, make sure the frontend code still works:
    - `WooCommerce > Extensions` (make sure you see the weekly deals notice (if you don't, run `localStorage.removeItem( 'wc-marketplaceNoticeClosed-promotion-0' )`))
    - `WooCommerce > Home` (make sure screen renders)
    - `Marketing > Overview` (make sure screen renders)
    - `Marketing > Coupons` (make sure screen renders)

#### New functionality

1. Apply [test.patch](https://github.com/user-attachments/files/17175811/test.patch) locally
2. Do `localStorage.setItem( 'debug', 'wc-admin:*' );` in your JS console to show Track events. Refresh the page, and events should show up in your JS console like this. (If you're using Chrome, enable verbose messages in the console.) 

<img width="500" alt="image" src="https://github.com/user-attachments/assets/9b08c013-5e5b-4dee-8fcd-c0ee8fe4be8c">

3. For each of `WooCommerce > Extensions`, `WooCommerce > Home`, `Marketing > Overview`, `Marketing > Coupons`:
    - Ensure you see the promo card for Black Friday. Track event `wcadmin_marketplace_promotion_viewed` should fire – check the props look OK.
    - Ensure link works. Track event `wcadmin_marketplace_promotion_actioned` should fire.
    - Ensure dismiss works (and doesn't show after reload). Track event `wcadmin_marketplace_promotion_dismissed` should fire.
    - To reset dismiss state for all of these four screens, run `localStorage.removeItem( 'wc-marketplaceDismissedPromos' )`

#### Bonus points: integration with WCCOM

1. Revert the patch you applied before.
2. Install [WP Crontrol](https://wordpress.org/plugins/wp-crontrol/) on your WooCommerce environment.
5. Set up your local WCCOM as in the test steps for 21881-gh-Automattic/woocommerce.com.
6. Apply this patch: [test-api.patch](https://github.com/user-attachments/files/17226716/test-api.patch). This points your WooCommerce environment at your local WCCOM.
7. Find and run the `woocommerce_marketplace_cron_fetch_promotions` cron event in `Tools > Cron Events` (example URL `http://localhost:8888/wp-admin/tools.php?page=crontrol_admin_manage_page&s=woocommerce_marketplace_cron_fetch_promotions`). This should update the `woocommerce_marketplace_promotions_v2` transient with the data from WCCOM.
8. Check that the promos from `New functionality` appear.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

bor0 committed:
Populate `window.wcMarketplace` for Marketing > Coupons

bor0 committed:
TEST CODE: Add promotions dummy data to the endpoint for now

See [this comment](https://github.com/Automattic/woocommerce.com/issues/21783#issuecomment-2376471712) as to how we plan to manage this through WCCOM

bor0 committed:
Initial implementation of the `<PromoCard>` component

This commit does the following:

- Implement the `PromoCard` component
- Refactor the existing `Promotions` component
    - Extract types
    - Add support for conditional rendering based on `pathname`
    - Accept optional parameter `promoCardOnly`
- Renders `PromoCard` wherever needed, as requested in the project thread.

Some remaining TODOs:

1. Dismiss functionality
2. Tracking
3. Figma design adjustments
    - CSS styling (Flexbox in Marketplace, etc.)
    - Percent image, wherever needed
    - Background image
4. BWC testing (e.g. ensuring that "Get more for your money" banner still works)

bor0 committed:
Make sure that WCCOM also returns path so that earlier versions of in-app marketplace don't break (path undefined error)

bor0 committed:
Update T&C link

After https://github.com/Automattic/woocommerce.com/issues/21840

bor0 committed:
Implement promo banner dismissal logic and tracks

andfinally committed:
Styling the WooCommerce > Extensions version of the promo component.

- Tweaked data on `class-wc-admin-assets.php` to pass `style` and `icon` attributes for that promo.
- Moved promo on this page from the marketplace `Header` component to the `Content`, above the Sales Banner notices.
- Deleted `percent.svg` and moved its content into a React component.
- CSS tweaks for this style of promo.
- Changes to elements of `PromoCard` component to allow the layout for this style.

andfinally committed:
Fixed some TypeScript issues.

- Added guard condition in case `promotion.id` is undefined.
- Returning null if it isn't - but below `useState` hook, so we conform to rules of hooks.
- Setting default value for `promotion.cta_label` to satisfy TypeScript, which expects a node.

Style tweaks to `.promo-cta`. Vertical centering of text, hover colour, margins.

bor0 committed:
Remove unnecessary unique identifier for each promo, and use a smarter way (URIs) to gather the visibility data

bor0 committed:
Lint

bor0 committed:
Revert test code

bor0 committed:
Changelog

bor0 commented on plugins/woocommerce-admin/client/marketplace/components/promotions/promotions.tsx:
```diff
@@ -12,37 +14,24 @@ declare global {
-
-const Promotions: () => null | JSX.Element = () => {
+const Promotions: ( {
+	promoCardOnly,
```
This feels like a dirty hack, but it's mainly there to preserve BWC. Ideally, each ad would use the API data format (and dismiss logic) as introduced here.

bor0 replied:
Addressed in 80078f8, much neater. Props @andfinally

bor0 committed:
Better code comments

andfinally committed:
Changed `promoCardOnly` attribute of `Promotions` to `format`, to allow for more types in future.

bor0 commented on plugins/woocommerce-admin/client/marketplace/components/promotions/promotions.tsx:
```diff
@@ -12,37 +14,21 @@ declare global {
+const Promotions: ( {
+	format,
+}: {
+	format?: string;
+} ) => null | JSX.Element = ( { format = 'notice' } ) => {
```
@andfinally, thank you for this commit! It's 💯.

My only proposal is to make the argument required.

```suggestion
	format: string;
} ) => null | JSX.Element = ( { format } ) => {
```

andfinally replied:
Thanks Boro, committed.

andfinally committed:
Update plugins/woocommerce-admin/client/marketplace/components/promotions/promotions.tsx

Co-authored-by: Boro Sitnikovski <buritomath@gmail.com>

andfinally committed:
Styling promo cards at different breakpoints.
Moved homescreen promo to after store management links.

andfinally committed:
Added condition to merge array of promos with the main `$promotions` array.

andfinally committed:
Added `useEffect` so we only record Track event for view of promotion once.
Changed Tracks prefix to the shorter `marketplace_promo_`.

andfinally committed:
Added Tracks events for both formats of promotions, `promo-card`, and the original `notice`.

andfinally committed:
Merging two style blocks.

andfinally committed:
Passing `format` to `recordEvent` instead of hard-coding it.

andfinally committed:
Addressing linter errors.

andfinally committed:
Moved operation merging promos into the main array into a separate method for greater clarity.

andfinally committed:
Moved Promotions component on WooCommerce > Home to below task list.

raja651 reviewed (commented):
@bor0  and @andfinally I can see the BCFM promo card on Home, extensions, marketing, coupons correctly. But i don't see the tracks event registering . May be i'm looking at the wrong place i checked the network tab on browser after applying this setting `localStorage.setItem( 'debug', 'wc-admin:*' );` according to test instructions.

![Screenshot 2024-10-03 at 2 51 23 PM](https://github.com/user-attachments/assets/31d2b3b2-512f-451f-89d1-7c853e1790a4)
![Screenshot 2024-10-03 at 2 50 45 PM](https://github.com/user-attachments/assets/16559ba0-e41b-4448-b9dd-0e2eab1f6324)

andfinally commented:
Thanks a lot @raja651, I'll have a look. I've just got a bunch of design feedback from Poli, will make some tweaks in the styling as well.

andfinally committed:
Styling tweaks.
Increased schedule of `woocommerce_marketplace_cron_fetch_promotions` cron job to twice daily.
Added guard conditions in `Promotions` in case `window.wcMarketplace.promotions` isn't defined or isn't an array.

andfinally commented:
Hi @raja651, just wondering, did you look in your JS console for the Tracks events? I've updated the test steps to clarify.

andfinally committed:
Reduced SVG size to 72px.

raja651 reviewed (approved):
Tracks are shown now, it was hidden on my console !

andfinally committed:
Fixing linter error.

andfinally committed:
Style tweaks. Using WP components colour vars for CTA and dismiss links.

andfinally reviewed (approved):
This looks and tests as expected.

andfinally committed:
Fixing linter error.

andfinally committed:
Reverting change to `schedule_cron_event`, as this method will not be present if only this branch is cherry-picked.
